KAFKA-14079 - Ack failed records in WorkerSourceTask when error tolerance is ALL

cshannon opened the pull request:
Make sure that if error tolerance is set to ALL that record failures are acked so the task will commit future record offsets properly and remove the records from internal tracking preventing a memory leak

cshannon committed:
KAFKA-14079 - Ack failed records in WorkerSourceTask when error tolerance is ALL

Make sure that if error tolerance is set to ALL that record failures
are acked so the task will commit future record offsets properly and remove
the records from internal tracking preventing a memory leak

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSourceTaskTest.java:
```diff
@@ -835,24 +836,41 @@ public void testSourceTaskIgnoresProducerException() throws Exception {
+        Whitebox.setInternalState(workerTask, "toSend", Arrays.asList(record1, record2, record3));
         Whitebox.invokeMethod(workerTask, "sendRecords");
 
+        final SubmittedRecords submittedRecords = Whitebox.getInternalState(workerTask, "submittedRecords");
+
+        //Make sure the unacked messages from the producer failure are 0 and have been cleared
+        int numUnackedMessages = Whitebox.getInternalState(submittedRecords, "numUnackedMessages");
+        assertEquals(0, numUnackedMessages);
+
+        //call commitableOffsets to verify that acked records will be cleared
+        //Before KAFKA-14079 acked records would not be cleared if the head was unacked due to an ignored record failure
+        submittedRecords.committableOffsets();
+
+        //Make sure size is 0 for all lists
+        for (Deque<SubmittedRecords.SubmittedRecord> deque : submittedRecords.records.values()) {
+            assertEquals(0, deque.size());
+        }
+
```
It's a little strange that we're interacting so heavily with a private field in this test, and it doesn't explicitly address the case that we're trying to test for (which is that we can continue to commit offsets after a record that the producer failed to send has been dropped).

What do you think about testing this case more directly? I'm thinking something like this might work:
- Using different source offsets for each of the records
- Invoking `workerTask.commitOffsets()`
- Verifying that the intended offsets were passed to the mocked offset writer via its `offset` method

C0urante reviewed (commented):
Thanks for identifying and fixing this bug @cshannon! I agree that this is worth getting into the upcoming 3.2.1 and 3.3.0 releases if possible; just have some thoughts about the testing strategy that hopefully shouldn't be too difficult to address.

I also saw your comments about the KIP-618 refactoring that make this change difficult to land on 3.3+; I'm happy to help out with that by reviewing or implementing a port of this fix onto that branch if you'd like--just let me know.

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java:
```diff
@@ -375,6 +375,10 @@ private boolean sendRecords() {
                                         preTransformRecord, e);
+
+                                //Make sure to clear out the record and update metrics
+                                submittedRecords.removeLastOccurrence(submittedRecord);
```
Actually, thinking about this a little more, should this record be removed, or acknowledged?

The reason we added the `SubmittedRecords::removeLastOccurrence` method is that we want to be able to re-submit the same record while handling retriable failures in `WorkerSourceTask::sendRecords`.

If we're not planning on re-submitting this record (which I don't believe we are), and we're trying to skip over the record permanently (which I believe we are), why not just allow its offsets to be committed? It's similar to how we handle offsets for records in aborted batches with exactly-once support: https://github.com/apache/kafka/blob/c020c94e043ca4a997797da91a226735803774ab/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java#L449-L450

On a per-record basis the impact is probably negligible, but if a large number of records fail to send and then the task is restarted, it'd save us some reprocessing of those same records.

cshannon replied:
Yeah I think you are right we can just ack instead of removing. It's a failed record so it wouldn't be retried, if should be skipped and it would be better performance as you mentioned. I can change that and also take a look at making test changes as well that you commented on.

cshannon commented:
@C0urante - Ok I went ahead and made the changes. I updated the description, changed the fix to ack the failed record and updated the tests to be simplified and to verify the last offset was committed properly now that the error record is acked and the internal tracking is empty.

I can go ahead and do a quick port to the trunk which I'll do now and submit another PR

cshannon commented:
@C0urante - I opened another PR that's slightly modified to fix trunk as well: #12415

C0urante reviewed (approved):
LGTM, thanks @cshannon

rhauch reviewed (approved):
Thanks, @cshannon, for reporting the problem, proposing a fix, and working with @C0urante toward approval. Thanks, @C0urante, for working quickly with @cshannon.

LGTM as a fix for the `3.2` branch and for inclusion in the upcoming 3.2.1.

cshannon commented:
@rhauch - Great, thanks for merging this so quickly. I've been running a custom build with this fix for a couple days and it has looked good so far so it will be nice to get 3.2.1 out soon.

rhauch commented:
As you know, the AK project is actively working on a 3.2.1 release; see the [Dev List thread](https://lists.apache.org/list?dev@kafka.apache.org:2022-7). I've responded to the thread mentioning this issue is no longer a blocker, and should appear in the forthcoming 3.2.1 release.
